File: 00986. Interval List Intersections/solution.js
/**
 * Problem: 986. Interval List Intersections
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Returns the intersection of two lists of closed intervals
 *
 * @param {number[][]} firstList - First list of intervals
 * @param {number[][]} secondList - Second list of intervals
 *
 * @returns {number[][]} - Intersection intervals
 */
const intervalIntersection = (firstList, secondList) => {
  // Array to store intersection intervals
  const intersections = []
  // Pointers for traversing both lists
  let firstIndex = 0,
    secondIndex = 0

  // Process both lists until we reach the end of either one
  while (firstIndex < firstList.length && secondIndex < secondList.length) {
    // Find potential intersection boundaries
    const startPoint = Math.max(
        firstList[firstIndex][0],
        secondList[secondIndex][0]
      ),
      endPoint = Math.min(firstList[firstIndex][1], secondList[secondIndex][1])

    // If valid intersection exists, add it to results
    if (startPoint <= endPoint) intersections.push([startPoint, endPoint])

    // Advance pointer of the interval that ends earlier
    if (firstList[firstIndex][1] < secondList[secondIndex][1]) firstIndex++
    else secondIndex++
  }

  return intersections
}

File: 00986. Interval List Intersections/solution.go
/**
 * Problem: 986. Interval List Intersections
 *
 * Difficulty: Medium
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func intervalIntersection(firstList [][]int, secondList [][]int) [][]int {
	// Initialize pointers for both lists
	firstIdx, secondIdx := 0, 0
	// Store the intersection intervals
	intersections := [][]int{}

	// Process both lists until we reach the end of either one
	for firstIdx < len(firstList) && secondIdx < len(secondList) {
		// Extract interval boundaries from both lists
		firstStart, firstEnd := firstList[firstIdx][0], firstList[firstIdx][1]
		secondStart, secondEnd := secondList[secondIdx][0], secondList[secondIdx][1]
		
		// Find potential intersection points
		maxStart := max(firstStart, secondStart)
		minEnd := min(firstEnd, secondEnd)

		// If valid intersection exists, add to result
		if maxStart <= minEnd {
			intersections = append(intersections, []int{maxStart, minEnd})
		}

		// Move pointer of the interval that ends earlier
		if firstEnd < secondEnd {
			firstIdx++
		} else {
			secondIdx++
		}
	}

	return intersections
}